crowbar_framework/app/models/service_object.rb

```diff
@@ -1902,8 +1941,9 @@ def skip_unready_nodes(bc, inst, new_elements, old_elements)
         pre_cached_nodes[n] ||= Node.find_by_name(n)
         node = pre_cached_nodes[n]
         next if node.nil?
-        # skip if nodes are on ready or crowbar_upgrade state, we dont need to do anything
-        next if ["ready", "crowbar_upgrade"].include?(node.state)
+        # skip if nodes are on ready, we dont need to do anything
+        next if node.state == "ready"
+        next if node.state == "crowbar_upgrade" && crowbar_upgrade_needs_to_run(node)
```

Comment:
Some comment on why this is actually needed would be helpful IMO. Remember how long it took you to understand the logic.